### PR TITLE
Fix infinite recursion in root detection code and lack of UE5 detection

### PR DESCRIPTION
**Pull request by sleeptightAnsiC**
TLDR - This PR fixes: https://github.com/adamrehn/ue4cli/issues/57 ,  https://github.com/adamrehn/ue4cli/issues/55 , https://github.com/adamrehn/ue4cli/issues/47 , https://github.com/adamrehn/ue4cli/pull/50 and few more things in the same code (see below)

Hey @adamrehn, I hope you are doing well!

In almost every UnrealManager... class there was a bug resulting in infinite recursion. It was caused by the usage of `getEngineVersion()` inside of `_detectEngineRoot()` which would never work - `getEngineVersion()` expects that Root is already known when said function is called, and the `_detectEngineRoot()` is only called when Root is NOT known, this leads to recursion in multiple cases. I just removed the `getEngineVersion()`  function call, it's not needed there at all. Additionally, `_detectEngineRoot()` was not working consistently across platforms, some of them were not detecting UE5. I fixed it, since it's the same code.

Let me know, if something is not right.

All changes done by this PR:
- remove `UnrealManagerBase::_getEngineRoot()`, it was only used in one place and hiding the code, split and move it's logic to the `UnrealManagerBase::getEngineRoot` and to the new function `UnrealManagerBase::getEngineRootOverride`
- move assertions from `_detectEngineRoot` overrides to `UnrealManagerBase::getEngineRoot()`. Some managers were not calling these assertions. Now each `_detectEngineRoot` will return `None` while failing to detect the Root which throws the same assertion shortly after. This is easier to maintain and probably makes more sense.
- fix: remove `_getEngineVersionDetails` function call from each `_detectEngineRoot` - it made no sense, was not needed and was causing the recursions. 
- fix: NameError in `UnrealManagerDarwin::_detectEngineRoot`
- fix: lack of UE5 detection on Windows in `UnrealManagerWindows::_detectEngineRoot`
- change assert message about root detection fail to be more meaningful: `'could not detect Unreal Engine root directory! Please, provide it manually via: ue4 setroot <ROOTDIR>'`
- add note to `UnrealManagerLinux::_detectEngineRoot` about possible bug with engine detection order 

Tested on ArchLinux with UE5.3.2 built from source, and on Win11 with UE5.3.2 downloaded via EpicGamesLauncher.
Sadly, I can't confirm if this works everywhere, there is too many of cases to test.

Cheers!

### CODE REVIEW

**Comment by adamrehn** [comment]
Hi @sleeptightAnsiC, thanks for contributing these fixes!

Having a look at the commit history, it appears that the infinite recursion bugs and macOS `NameError` issue were introduced when #44 was merged. Although I can no longer remember the details, I suspect that I might have merged that pull request without actually testing it myself, likely due to the increasing constraints on my time that subsequently led to my complete unresponsiveness to issues on the ue4cli and conan-ue4cli repositories over the past couple of years.

In the interests of not repeating that mistake, I've tested your changes across Linux, macOS and Windows. Engine root path detection appears to be functioning as expected across all three operating systems, and the only minor inconsistency I've found is that `UnrealManagerLinux._detectEngineRoot()` still throws an exception, rather than returning `None` as the updated logic for Windows and macOS now do. I'll fix this after merging the existing changes, to ensure consistent behaviour across all platforms.

**Comment by sleeptightAnsiC** [comment]
Thanks! and sorry for missing the last assertion
